UNL-CSE-AMBASSADORS/VITA
# 231 Simple International Station Limiting

This PR enables the functionality of limiting the number of appointments happening at a "station" at the international tax site. Basically, if the appointment is an international appointment, we now collect the country the person is from so we can determine if it is a China, India, Treaty, or Non-Treaty appointment. The limits currently for each "station" are hard-coded on the back-end. I wasn't sure a better approach at this time, and just wanted to get something up that works.  

Note that this PR isn't as bad as it looks line-of-code-wise, I added the `package-lock.json`--that's the majority of this PR.

**Notable Changes:**
* The appointment picker now uses a string instead of the `studentScholar` boolean--this `appointmentType` string takes on the values of "residential", "china", "india", "treaty", or "non-treaty". 


**Miscellaneous Fixes in this PR**
* Fixed a small bug in self-service appointment rescheduling that prevented a user that didn't have an email from being able to validate their information
* Appointment rescheduling now defaults to whatever the appointment was originally created as (i.e. if it's a china, the appointment picker will default show china appointments instead of residential). This fix also enables international appointments to actually reschedule for another international appointment--they couldn't do this before, whoops!
* Removed some dead code in the queue that was referencing old dependents functionality
* Updated some NPM packages because I updated Node/NPM, and I had to update these packages so that `npm run watch`
* Include the `package-lock.json` in the project--this is the proper thing to do, and we haven't been doing it this whole time
* Update dates on index page
* Update date for international appointment sign-ups

## Testing
It's a doozy, please bear with me. Note that you may need to run `npm install` again or `npm upgrade` since I changed the versions of our packages. 
1. Run the DB scripts and the build script
2. Change `$dateAppointmentSignUpsStart` variable in `/signup/signup.php` to be `Y-01-01`.
3. Navigate to /signup
4. Go on the path to create an international appointment
    * You should eventually see a drop-down asking you to specify which country you're a citizen or resident of
        * If you choose China, the appointment picker will update for China appointments
        * If you choose India, the appointment picker will update for India appointments
        * If you choose Other, the appointment picker will update for Non-Treaty appointments
        * If you choose any other country, the appointment picker will update for Treaty appointments
    * If you change the visa/length of time you've been in the country, the country thing should go away and reset
5. Sign up for a China appointment
    * Open the self-service reschedule link in a new tab, we will test that later on
6. Navigate to /signup again
7. Go on the path for another China appointment
    * Go to the network tab in chrome and note that in the results from the `getTime` query, there is now one less spot for China appointments (default there is 15, the test DB script puts 1 in, and you just signed up for one, so it should say there are 13 slots).
8. Sign in a preparer@test.test
9. Go to the appointment management page
10. Search for the appointment you just created and click on it
    * You should see "Tax Type" as a new field in the information, and it should say "china"
11. Scroll to the reschedule functionality at the bottom
    * You should see buttons that enable you to change the "Tax Type", and it should say that it is currently "china"
    * Hitting any of the options should update the appointment picker
    * Reschedule the appointment for a residential appointment
        * It should be successful, and the information at the top should now say "residential"
        * Refresh the page and search for that appointment again, click on it, and the information should still be correct
12. Go to the self-service reschedule tab
    * Input the proper information
13. The appointment picker should have defaulted to China appointments
    * You can verify by checking the network tab


## Additional Testing
Sorry I touched random crap in this one, so there is a few other things to test to make sure they still work
* Self-service appointment reschedule for
    * A client without an email
        * Ensure you can validate information successfully
    * A client with an email
        * Ensure you can validate information successfully
    * Ensure that invalid information doesn't work

## Database Evolutions
The follow SQL script will need to be ran before this is deployed. This simply adds the new question / answers required for this functionality.

```sql
INSERT INTO Question (text, lookupName)
	VALUES ("What country are you a citizen or resident of?", "treaty_type");

INSERT INTO PossibleAnswer (text)
	VALUES ("china");

INSERT INTO PossibleAnswer (text)
	VALUES ("india");

INSERT INTO PossibleAnswer (text)
	VALUES ("treaty");

INSERT INTO PossibleAnswer (text)
	VALUES ("non-treaty");
```
